staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			continue
 		}
 		if evalResult != types.True {
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if compiled.MessageExpression != nil {
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
+				if msgErr != nil {
+					if msgErr.errType == runtimeCostErr {
+						errs = append(errs, field.InternalError(fldPath, msgErr.err))
+						return errs, -1
+					} else {
+						klog.Warning(msgErr.err)
+						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+						remainingBudget = newRemainingBudget
+					}
+				} else {
+					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					remainingBudget = newRemainingBudget
+				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, fmt.Sprintf("failed rule: %s", ruleErrorString(rule))))
+				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
 			}
 		}
 	}
 	return errs, remainingBudget
 }
 
+type _errType int
+
+const (
+	runtimeCostErr _errType = iota
+	evalErr
+)
+
+type messageExpressionErr struct {
+	errType _errType
+	err     error
+}
+
+// evalMessageExpression takes a context, Program and source code representing the MessageExpression, activation,
+// and remaining, runtime cost budget, returns the evaluated message and the remaining budget, or an error if one
+// occurred during evaluation.
+func evalMessageExpression(ctx context.Context, expr celgo.Program, exprSrc string, activation interpreter.Activation, remainingBudget int64) (string, int64, *messageExpressionErr) {
+	evalResult, evalDetails, err := expr.ContextEval(ctx, activation)
+	if evalDetails == nil {
+		return "", -1, &messageExpressionErr{
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression: %q", exprSrc),
+		}
+	}
+	rtCost := evalDetails.ActualCost()
+	if rtCost == nil {
+		return "", -1, &messageExpressionErr{
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression: %q", exprSrc),
+		}
+	} else {
+		if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
+			return "", -1, &messageExpressionErr{
+				runtimeCostErr,
+				errors.New("messageExpression evaluation failed due to running out of cost budget, no further validation rules will be run"),
+			}
+		}
+	}
+	if err != nil {
```

Comment:
Differentiate the cost exceed err from other evalErr?

Reply:
If the cost is exceeded we'll get a `runtimeCostErr`:

```golang
if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
	return "", -1, &messageExpressionErr{
		runtimeCostErr,
		errors.New("messageExpression evaluation failed due to running out of cost budget, no further validation rules will be run"),
	}
}
```

That's different from `evalErr` - did you mean differentiate the runtime cost error from other `runtimeCostErr`s? I feel the current set of `runtimeCostErr`s is a pretty reasonable group, since I don't think we want to continue if either a) `evalDetails` is `nil` b) `evalDetails.ActualCost()` is nil c) we're out of budget.

Reply:
This is cost overall budget exceed. I am talking about the per expression limit exceed^^

Reply:
Ref: https://github.com/kubernetes/kubernetes/blob/a5f7d533b08e05dbf0ce23db4bdd62ad9df64a62/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go#L247-L249

Reply:
Gotcha - marking this as resolved once I add a branch + tests.

Reply:
(marking as resolved due to 8ebb12c6b8efb2f517e09ec3b676b7cfbb0f3181)